dvc/data/stage.py

```diff
@@ -247,6 +247,24 @@ def _get_staging(odb: "ObjectDB") -> "ObjectDB":
     return ReferenceObjectDB(fs, fs_path, state=state)
 
 
+def _load_untyped_dir_hash(odb, fs, fs_path, hash_info, meta, staging):
+    from .tree import Tree
+
+    obj = Tree.load(odb, hash_info.as_raw())
+    obj.hash_info = hash_info
+    meta.nfiles = len(obj)
```

Comment:
Seems like this is duplicating the older logic (which is minor), but also note that you are not validating the object anymore (see check() call above). So maybe this is a hint that you need to look into somehow unifying what you have here with what has already been in `_load_from_state`.